blivet/zfcp.py

```diff
@@ -47,62 +49,218 @@ def logged_write_line_to_file(fn, value):
             util.run_program(["zfcp_cio_free", "-d", self.devnum])
 
-        if not os.path.exists(online):
+        if not os.path.exists(self._device_online_path):
             raise ValueError(_("zFCP device %s not found, not even in device ignore list.") %
                              (self.devnum,))
 
+    def _set_zfcp_device_online(self):
+        """Set the zFCP device online.
+
+        :raises: ValueError if the device cannot be set online
+        """
+
         try:
-            f = open(online, "r")
-            devonline = f.readline().strip()
-            f.close()
+            with open(self._device_online_path) as f:
+                devonline = f.readline().strip()
             if devonline != "1":
-                logged_write_line_to_file(online, "1")
+                logged_write_line_to_file(self._device_online_path, "1")
         except OSError as e:
             raise ValueError(_("Could not set zFCP device %(devnum)s "
                                "online (%(e)s).")
                              % {'devnum': self.devnum, 'e': e})
 
+    def _set_zfcp_device_offline(self):
+        """Set the zFCP device offline.
+
+        :raises: ValueError if the device cannot be set offline
+        """
+
+        try:
+            logged_write_line_to_file(self._device_online_path, "0")
+        except OSError as e:
+            raise ValueError(_("Could not set zFCP device %(devnum)s "
+                               "offline (%(e)s).")
+                             % {'devnum': self.devnum, 'e': e})
+
+    def _is_scsi_associated_with_fcp(self, fcphbasysfs, _fcpwwpnsysfs, _fcplunsysfs):
+        """Decide if the SCSI device with the provided SCSI attributes
+        corresponds to the zFCP device.
+
+        :returns: True or False
+        """
+
+        return fcphbasysfs == self.devnum
+
+    def online_device(self):
+        """Initialize the device and make its storage block device(s) ready to use.
+
+        :returns: True if success
+        :raises: ValueError if the device cannot be initialized
+        """
+
+        self._free_device()
+        self._set_zfcp_device_online()
+        return True
+
+    def offline_scsi_device(self):
+        """Find SCSI devices associated to the zFCP device and remove them from the system."""
+
+        # A list of existing SCSI devices in format Host:Bus:Target:Lun
+        scsi_devices = [f for f in os.listdir(scsidevsysfs) if re.search(r'^[0-9]+:[0-9]+:[0-9]+:[0-9]+$', f)]
+
+        scsi_device_found = False
+        for scsidev in scsi_devices:
+            fcpsysfs = os.path.join(scsidevsysfs, scsidev)
+
+            with open(os.path.join(fcpsysfs, "hba_id")) as f:
+                fcphbasysfs = f.readline().strip()
+            with open(os.path.join(fcpsysfs, "wwpn")) as f:
+                fcpwwpnsysfs = f.readline().strip()
+            with open(os.path.join(fcpsysfs, "fcp_lun")) as f:
+                fcplunsysfs = f.readline().strip()
```

Comment:
FYI.
Since this is only updated old code, it's OK to keep deprecated and zfcp-specific sysfs attributes rather than using common code mechanisms, which I would prefer for new code.
https://www.ibm.com/docs/en/linux-on-systems?topic=devices-mapping-sysfs-representations